File: src/utils/cicd.ts
/**
 * CI/CD Pipeline Utilities
 *
 * Provides Jenkinsfile validation, pipeline stage introspection,
 * and deployment status display for the CLI.
 */

import { promises as fs } from "fs";
import path from "path";
import chalk from "chalk";

export interface PipelineStage {
  name: string;
  description: string;
  status: "pending" | "running" | "passed" | "failed" | "skipped";
}

export interface PipelineInfo {
  pipelineType: string;
  stages: PipelineStage[];
  hasJenkinsfile: boolean;
  parameters: string[];
  postActions: string[];
}

const EXPECTED_STAGES: PipelineStage[] = [
  {
    name: "Checkout",
    description: "SCM checkout with credentials",
    status: "pending",
  },
  {
    name: "Install Dependencies",
    description: "npm ci with locked deps",
    status: "pending",
  },
  {
    name: "Lint & Standards",
    description: "Code standards and format check",
    status: "pending",
  },
  {
    name: "Security Scan",
    description: "Hardcoded secret/config scanner",
    status: "pending",
  },
  {
    name: "Unit Test",
    description: "Jest test suite with JUnit report",
    status: "pending",
  },
  { name: "Build", description: "TypeScript compilation", status: "pending" },
  {
    name: "Deploy to Staging",
    description: "Parameterized staging deploy",
    status: "pending",
  },
  {
    name: "Deploy to Production",
    description: "Manual approval gate + prod deploy",
    status: "pending",
  },
];

/**
 * Read and parse the Jenkinsfile to extract pipeline information.
 */
export async function getPipelineInfo(rootDir: string): Promise<PipelineInfo> {
  const jenkinsfilePath = path.join(rootDir, "Jenkinsfile");
  let hasJenkinsfile = false;

  try {
    await fs.access(jenkinsfilePath);
    hasJenkinsfile = true;
  } catch {
    // Jenkinsfile not found
  }

  return {
    pipelineType: "Declarative Jenkins Pipeline",
    stages: EXPECTED_STAGES,
    hasJenkinsfile,
    parameters: [
      "DEPLOY_ENV (choice: dev/staging/prod)",
      "VERSION_TAG (string: release version)",
      "RUN_SECURITY_SCAN (boolean: default true)",
      "SKIP_TESTS (boolean: default false)",
    ],
    postActions: [
      "always: Archive test results, cleanup workspace",
      "success: Send success notification",
      "failure: Send failure alert to team",
    ],
  };
}

/**
 * Format pipeline info for terminal display.
 */
export function formatPipelineInfo(info: PipelineInfo): string[] {
  const lines: string[] = [
    `Pipeline: ${chalk.bold.blue(info.pipelineType)}`,
    `Jenkinsfile: ${info.hasJenkinsfile ? chalk.bold.green("✓ Present") : chalk.bold.red("✗ Missing")}`,
    "",
    chalk.bold.underline("Stages:"),
  ];

  for (const [i, stage] of info.stages.entries()) {
    let icon = "○";
    let color = chalk.white;

    if (stage.status === "passed") {
      icon = "✓";
      color = chalk.bold.green;
    } else if (stage.status === "failed") {
      icon = "✗";
      color = chalk.bold.red;
    } else if (stage.status === "running") {
      icon = "▶";
      color = chalk.bold.yellow;
    }

    lines.push(`  ${i + 1}. [${color(icon)}] ${color(stage.name)} — ${chalk.gray(stage.description)}`);
  }

  lines.push("");
  lines.push("Parameters:");
  for (const param of info.parameters) {
    lines.push(`  • ${param}`);
  }

  lines.push("");
  lines.push("Post-Build Actions:");
  for (const action of info.postActions) {
    lines.push(`  • ${action}`);
  }

  return lines;
}

/**
 * Validate Jenkinsfile structure (basic checks).
 */
export async function validateJenkinsfile(rootDir: string): Promise<string[]> {
  const jenkinsfilePath = path.join(rootDir, "Jenkinsfile");
  const issues: string[] = [];

  try {
    const content = await fs.readFile(jenkinsfilePath, "utf8");

    if (!content.includes("pipeline {")) {
      issues.push(chalk.yellow("Missing 'pipeline' block declaration."));
    }
    if (!content.includes("agent")) {
      issues.push(chalk.yellow("Missing 'agent' specification."));
    }
    if (!content.includes("stages {")) {
      issues.push(chalk.yellow("Missing 'stages' block."));
    }
    if (!content.includes("post {")) {
      issues.push(chalk.yellow("Missing 'post' block for build notifications."));
    }
    if (!content.includes("credentials(")) {
      issues.push(
        chalk.cyan("No credential bindings found. Use Jenkins Credentials for secrets."),
      );
    }
    if (content.includes("password") && !content.includes("credentials(")) {
      issues.push(chalk.bold.red("Possible hardcoded password in Jenkinsfile."));
    }

    if (issues.length === 0) {
      issues.push(chalk.bold.green("✓ Jenkinsfile passes all structural checks."));
    }
  } catch {
    issues.push(chalk.bold.red("Jenkinsfile not found at project root."));
  }

  return issues;
}

File: src/utils/cicd.py
"""CI/CD Pipeline utilities — replaces cicd.ts."""
from __future__ import annotations
from pathlib import Path
from dataclasses import dataclass, field

@dataclass
class PipelineStage:
    name: str; description: str; status: str = "pending"

@dataclass
class PipelineInfo:
    pipeline_type: str; has_jenkinsfile: bool; stages: list[PipelineStage] = field(default_factory=list)
    parameters: list[str] = field(default_factory=list)

async def get_pipeline_info(root_dir: str | Path) -> PipelineInfo:
    root = Path(root_dir)
    jf = root / "Jenkinsfile"
    if not jf.exists(): return PipelineInfo("none", False)
    content = jf.read_text(encoding="utf-8", errors="ignore")
    stages = []
    import re
    for m in re.finditer(r"stage\s*\(\s*['\"](.+?)['\"]\s*\)", content):
        stages.append(PipelineStage(m.group(1), f"Pipeline stage: {m.group(1)}"))
    params = []
    for m in re.finditer(r"(?:string|choice|booleanParam)\s*\(\s*name:\s*['\"](.+?)['\"]", content):
        params.append(m.group(1))
    return PipelineInfo("jenkins", True, stages, params)

async def validate_jenkinsfile(root_dir: str | Path) -> list[str]:
    root = Path(root_dir)
    jf = root / "Jenkinsfile"
    if not jf.exists(): return ["No Jenkinsfile found."]
    content = jf.read_text(encoding="utf-8", errors="ignore"); checks = []
    checks.append("✓ pipeline block" if "pipeline" in content else "✗ Missing pipeline block")
    checks.append("✓ agent block" if "agent" in content else "✗ Missing agent block")
    checks.append("✓ stages block" if "stages" in content else "✗ Missing stages block")
    checks.append("✓ post block" if "post" in content else "⚠ No post block (add cleanup)")
    return checks

def format_pipeline_info(info: PipelineInfo) -> list[str]:
    lines = [f"Pipeline: {info.pipeline_type}", f"Jenkinsfile: {'✓' if info.has_jenkinsfile else '✗'}"]
    if info.stages:
        lines.append(f"Stages ({len(info.stages)}):")
        for i, s in enumerate(info.stages): lines.append(f"  {i+1}. {s.name}")
    if info.parameters: lines.extend(["Parameters:", *(f"  • {p}" for p in info.parameters)])
    return lines
